Add amcl3 demo

pvela2017 opened the pull request:
### Proposed changes

Added a demo for the AMCL3 algorithm based on the likelihood sensor model 3D (https://github.com/Ekumen-OS/beluga/pull/440)

#### Type of change

- [ ] 🐛 Bugfix (change which fixes an issue)
- [X] 🚀 Feature (change which adds functionality)
- [ ] 📚 Documentation (change which fixes or extends documentation)

### Checklist

_Put an `x` in the boxes that apply. This is simply a reminder of what we will require before merging your code._

- [ ] Lint and unit tests (if any) pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [X] All commits have been signed for [DCO](https://developercertificate.org/)

### Additional comments

hidmic commented on localization/beluga_demo_amcl3_localization/src/beluga_amcl3_demo.cpp:
```diff
@@ -0,0 +1,466 @@
+        point_color.b = 1.0;
+        point_color.a = 1.0;
+
+        for (openvdb::FloatGrid::ValueOnCIter iter = grid->cbeginValueOn(); iter; ++iter) {
```
@pvela2017 the condition expression is unusual, is there no `cendValueOn()`?

pvela2017 replied:
As discussed in the meeting it is ok.

> OpenVDB iterators are not STL-compatible in that one can always request an iterator that points to the beginning of a collection of elements (nodes, voxels, etc.), but one usually cannot request an iterator that points to the end of the collection. (This is because finding the end might require a full tree traversal.) 

[Reference](https://www.openvdb.org/documentation/doxygen/overview.html#treeNodeIterRef)

hidmic commented on localization/beluga_demo_amcl3_localization/src/beluga_amcl3_demo.cpp:
```diff
@@ -0,0 +1,466 @@
+        visualization_msgs::msg::Marker marker_msg;
+        std_msgs::msg::ColorRGBA point_color;
+        openvdb::CoordBBox bbox = grid->evalActiveVoxelBoundingBox();
+        double min_z, max_z;
```
@pvela2017 nit: better define and assign on the same line, then it can be `const`

pvela2017 replied:
Changed in 64e5615601d05c975eb50e7cbd60ce2acc80b3b4

hidmic commented on localization/beluga_demo_amcl3_localization/src/beluga_amcl3_demo.cpp:
```diff
@@ -0,0 +1,466 @@
+            marker_msg.colors.push_back(point_color);
+        }
+
+        double size                   = grid->transform().voxelSize()[0];
```
@pvela2017 nit: can be `const`

pvela2017 replied:
Changed in 64e5615601d05c975eb50e7cbd60ce2acc80b3b4

hidmic commented on docker/run.sh:
```diff
@@ -66,10 +66,31 @@ while [[ "$1" != "" ]]; do
+                     "localization/beluga_demo_bearing_localization" 
+                     "localization/beluga_demo_fiducial_localization" 
+                     "localization/beluga_demo_lidar_localization"
+    )
```
@pvela2017 I'm a bit ambivalent about this, but I can't think of a better solution. CC @glpuga for thoughts.

glpuga replied:
I don't like it. I'm assuming these don't build or have missing dependencies in Jazzy.

Why supporting anything other than Humble, if this is a demos repository and we provide a docker to run it?

glpuga replied:
Oh, I remember now.

> Jazzy is new LTS distro. Having some our demos be incompatible with the latest LTS distro of ROS 2 is not good.

I'm not so taxative. People are still by far mostly using Humble, while Jazzy is at best "up-and-coming" right now. Core stuff should support both already, but leaf-stuff -like this self-contained demos repo- does not need to.

In any case, there's no reason I can think off the top of my head why these would not build in Jazzy. Is there any external dependency not available? 

Looking at the package.xml files, everything seems either standard packages or packages with code we wrote ourselves. Indirect dependencies might be gazebo (classic?) and [apriltag_ros](https://github.com/Adlink-ROS/apriltag_ros) which we use the "foxy-devel" branch (because there's no other).

hidmic replied:
> In any case, there's no reason I can think off the top of my head why these would not build in Jazzy. Is there any external dependency not available?

@pvela2017 ? I think it was Gazebo related. Isn't Gazebo Classic gone?

hidmic replied:
I'll give this a shot. Maybe I can get all demos working on Jazzy. Then the problem goes away.

hidmic replied:
So, https://github.com/ROBOTIS-GIT/turtlebot3_simulations has not seen a release since Humble and it doesn't look like it will ever do. Nav2 folks spinned up https://github.com/ros-navigation/nav2_minimal_turtlebot_simulation in place for it. Some of our demos already depend on Nav2 code, so I'll go ahead and use this for both Humble and Jazzy. It's source distributed anyways.

hidmic replied:
> Should we update everything to Gazebo ignition?

@pvela2017 indeed. I'm on it.

pvela2017 replied:
> > Should we update everything to Gazebo ignition?
> 
> @pvela2017 indeed. I'm on it.

Can I help?

hidmic replied:
Sure! I'll tag you on the PR as soon as it is up for review. If you can then adjust this PR for it, that'd be amazing.

hidmic reviewed (commented):
First pass

pvela2017 commented on localization/beluga_demo_amcl3_localization/launch/demo_botanic_garden_amcl3_localization.launch.py:
```diff
@@ -0,0 +1,131 @@
+    ]
+
+    return LaunchDescription(nodes)
+    
```
Changed in b15a8cc4bd89b13c6a541af279fba826c76f6856 and 471d7167cb6340669907a80537c3f3b7141e0d52

pvela2017 commented on localization/beluga_demo_amcl3_localization/scripts/marker_publisher.py:
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3
+
```
I like the line type visualization

pvela2017 commented on localization/beluga_demo_amcl3_localization/scripts/tf_publisher.py:
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python3
+
```
I will try

pvela2017 replied:
I tried to implement this on an other project. But there when playing the bag there was a warning about the timestamp. I dont know if it is a bug or I did something wrong. I will try again for this project

pvela2017 replied:
Its working! How can I send you the new rosbag?
Updated in 5978183186f42e5b9b42779e6863a6ef131dd51c

hidmic replied:
@pvela2017 GDrive share?

pvela2017 commented on localization/beluga_demo_amcl3_localization/src/beluga_amcl3_demo.cpp:
```diff
@@ -0,0 +1,466 @@
+
+class Amcl3 {
+ public:
+    explicit Amcl3(motion_model &&motion, sensor_model &&sensor, Sophus::SE3d pose, Sophus::Matrix6d covariance, execution_policy policy = std::execution::par)
```
Changed in 64e5615601d05c975eb50e7cbd60ce2acc80b3b4

pvela2017 commented on localization/beluga_demo_amcl3_localization/src/beluga_amcl3_demo.cpp:
```diff
@@ -0,0 +1,466 @@
+            beluga::actions::propagate(execution_policy_, motion_model_(control_action_window_ << base_pose_in_odom)) | 
+            beluga::actions::reweight(execution_policy_, sensor_model_(std::move(pointcloud))) |                        
+            beluga::actions::normalize(execution_policy_) |
+            beluga::actions::assign;
```
Changed in 64e5615601d05c975eb50e7cbd60ce2acc80b3b4

pvela2017 commented on localization/beluga_demo_amcl3_localization/src/beluga_amcl3_demo.cpp:
```diff
@@ -0,0 +1,466 @@
+
+        // Initial position
+        const auto pose = Sophus::SE3d{
+            Sophus::SO3d{/*kInitialPoseYaw, kInitialPosePitch, kInitialPoseRoll*/},
```
Changed in 64e5615601d05c975eb50e7cbd60ce2acc80b3b4

pvela2017 commented on docker/images/jazzy/Dockerfile:
```diff
@@ -0,0 +1,107 @@
+    libopenvdb-dev \
+    python3-pip \
+    mc \
+    ros-jazzy-pcl-ros \
```
Changed in 150be6eeed2f7ea0503413988c14e724c697db7c

hidmic commented:
@pvela2017 https://github.com/Ekumen-OS/beluga/pull/440 got merged. Are you still up for wrapping this up?

pvela2017 commented:
> @pvela2017 [Ekumen-OS/beluga#440](https://github.com/Ekumen-OS/beluga/pull/440) got merged. Are you still up for wrapping this up?

Yep! I'm here

hidmic commented on docker/files/jazzy_base.repos:
```diff
@@ -0,0 +1,5 @@
+repositories:
+  beluga:
+    type: git
+    url: https://github.com/pvela2017/beluga.git
```
This can now point to the main repository.

pvela2017 replied:
Changed in 623f842b977cb8e07618f9d08e8d0756b6940778

hidmic reviewed (approved):
Great job. Left a couple final comments but I'm overall satisfied with this PR.

pvela2017 commented on localization/beluga_demo_amcl3_localization/src/beluga_amcl3_demo.cpp:
```diff
@@ -0,0 +1,466 @@
+    const auto& particles() const { return particles_; }
+
+    private:
+    beluga::TupleVector<particle_type> particles_;
```
Fixed

pvela2017 commented on localization/beluga_demo_amcl3_localization/src/beluga_amcl3_demo.cpp:
```diff
@@ -0,0 +1,517 @@
+  const std::string kOdomFrameId = "odom";
+  const std::string kMapFile =
+      "/home/developer/ws/src/beluga_demo/localization/"
+      "beluga_demo_amcl3_localization/maps/map.vdb";
```
Updated in 8d7f101f01eee3b327b60e15bfcc2679814d749f

hidmic commented:
@pvela2017 this is a ready for a rebase. No need for the weird bind-mounts anymore.

pvela2017 commented:
> @pvela2017 this is a ready for a rebase. No need for the weird bind-mounts anymore.

Rebased. I modified the aliases to not compile anything related with `vdb` if it is no Jazzy and added the new demo on the `README`.

hidmic commented on docker/files/DOTaliases:
```diff
@@ -1,6 +1,12 @@
+  alias demo_build='cd ~/ws \
+    && colcon build --symlink-install --packages-up-to-regex beluga_demo --cmake-args -DCMAKE_BUILD_TYPE=Release \
+    && source install/setup.bash'
+fi
```
@pvela2017 meta: if you are up for https://github.com/Ekumen-OS/beluga/issues/474  once we've sorted out the OpenVDB dependency in Humble, we should come back here and rollback the distro specific switches.

pvela2017 replied:
Fixed on [61f59b2](https://github.com/Ekumen-OS/beluga-demos/pull/9/commits/61f59b2c7dd196ebfc0a8b583c10240446b12c45)

hidmic commented on localization/beluga_demo_amcl3_localization/package.xml:
```diff
@@ -0,0 +1,24 @@
+  <name>beluga_demo_amcl3_localization</name>
+  <version>1.0.0</version>
+  <description>Demos of Beluga AMCL 3D based on a botanic garden dataset.</description>
+  <maintainer email="glpuga@ekumenlabs.com">Gerardo Puga</maintainer>
```
@pvela2017 meta: you should add yourself as `<author/>` here too

pvela2017 replied:
Updated in [10c7710](https://github.com/Ekumen-OS/beluga-demos/pull/9/commits/10c7710f8d77a5bb252a67f8f341037adcba64bb)

hidmic commented on localization/beluga_demo_amcl3_localization/scripts/marker_publisher.py:
```diff
@@ -0,0 +1,120 @@
+        point.x = position.x
+        point.y = position.y
+        point.z = position.z
+        self.amcl3_points.append(point)
```
@pvela2017 hmm, won't this grow unbounded over time? What's the intent?

pvela2017 replied:
The purpose is to have a visual comparison of the whole path, like this:

![Screenshot from 2025-02-07 14-44-35](https://github.com/user-attachments/assets/11d6edcd-f7ed-494e-bf52-de201e6ab059)

hidmic replied:
Understood, nevermind.

hidmic commented on localization/beluga_demo_amcl3_localization/rosbags/botanic_garden_1005_07/metadata.yaml:
```diff
@@ -0,0 +1,43 @@
+rosbag2_bagfile_information:
```
@pvela2017 hmm, why have datasets' metadata here? Bags in ROS 2 are typically a `metadata.yaml` file and a vendor specific storage file (e.g. an SQLite file DB, an MCAP file, etc.). They come in pairs. Why not keep it that way?

pvela2017 replied:
The storage files are too big, same with the maps. I think you are right, I think I should include the whole rosbag folder into the `.gitignore` file and add instructions about how to download them.

What do you reckon?

hidmic replied:
> I think I should include the whole rosbag folder into the .gitignore file and add instructions about how to download them.

I fully agree with that.

pvela2017 replied:
Updated in [5d10c07](https://github.com/Ekumen-OS/beluga-demos/pull/9/commits/5d10c078a5005b78aba3f260cc8fb7db99891ef5)

hidmic commented on README.md:
```diff
@@ -70,15 +71,16 @@ You can close the demo by pressing `Ctrl+C` in the terminal where you initially
+| `apriltags_localization_demo`        | Humble/Jazzy | Simple custom localization node using the `beluga` library to localize the robot within a large $10m \times 10m$ area using Apriltag markers as landmarks. The code of this localization node can be found within this repository [here](https://github.com/Ekumen-OS/beluga-demos/blob/main/localization/beluga_demo_fiducial_localization/src/beluga_ ...
+| `light_beacons_localization_demo`    | Humble/Jazzy | Simple custom localization node using the `beluga` library to localize the robot within a large $10m \times 10m$ area using light sources as landmarks. The code of this localization node can be found within this repository [here](https://github.com/Ekumen-OS/beluga-demos/blob/main/localization/beluga_demo_bearing_localization/src/beluga_bmcl ...
+| `nav2_integration_demo`              | Humble/Jazzy | Demo using the [_beluga_amcl_](https://github.com/Ekumen-OS/beluga/tree/main/beluga_amcl) node in lieu of the [_nav2_amcl_](https://github.com/ros-navigation/navigation2/tree/main/nav2_amcl) in a [Nav2](https://docs.nav2.org/) stack to navigate around a large office cluttered with unmapped obstacles.
+| `amcl3_localization_demo`              | Jazzy | Custom 3D localization node using the [_beluga_vdb_](https://github.com/Ekumen-OS/beluga/tree/main/beluga_vdb) library to localize the robot in a botanic garden [dataset](https://github.com/robot-pesg/BotanicGarden). The code of this localization node is based on `OpenVDB` and can be found within this repository [here](https://github.com/Ekumen-O ...
```
@pvela2017 meta: unlike other demos in this repository, running the `amcl3_localization_demo` requires separate dataset provisioning (right?). We should perhaps mention that.

pvela2017 replied:
Should I write the instructions on a `README` inside the demo folder and link it here?

hidmic replied:
That sounds appropriate, yes,

pvela2017 replied:
Updated on [5d10c07](https://github.com/Ekumen-OS/beluga-demos/pull/9/commits/5d10c078a5005b78aba3f260cc8fb7db99891ef5)

How should we suggest to download the maps and rosbag files, currently I just wrote on the [`README`](https://github.com/Ekumen-OS/beluga-demos/blob/5d10c078a5005b78aba3f260cc8fb7db99891ef5/localization/beluga_demo_amcl3_localization/doc/README.md)

```bash
wget url && \
tar -xf file.tar.gz
```

hidmic reviewed (commented):
Great job @pvela2017, as always. Tagging @nahueespinosa and @glpuga for feedback too.

pvela2017 commented:
@hidmic 
Here are the new rosbags 
https://drive.google.com/drive/folders/1BAa4U35xnPMPY7RjMHsCF_-uWxES5HHB?usp=sharing

hidmic commented on localization/beluga_demo_amcl3_localization/doc/README.md:
```diff
@@ -0,0 +1,95 @@
+# Beluga VDB: AMCL3 Localization Demo!
+
+This package provides 2 maps based on the [botanic garden dataset](https://github.com/robot-pesg/BotanicGarden). The rosbags from the botanic garden dataset were modified (`/odom` topic was added and several topics were removed) to be able to run on this demo. Currently, only 2 sequences can be tested 1005-07 and 1006-01.
```
@pvela2017 nit: 
```suggestion
This package provides 2 maps based on the [botanic garden dataset](https://github.com/robot-pesg/BotanicGarden). The rosbags from the botanic garden dataset were modified (`/odom` topic was added and several topics were removed) to be able to run on this demo. Currently, only 2 sequences can be tested: 1005-07 and 1006-01.
```

pvela2017 replied:
Updated  on [a4e9f6b](https://github.com/Ekumen-OS/beluga-demos/pull/9/commits/a4e9f6b9c8399f41b0280c8c38c07443acad3a15)

hidmic commented on localization/beluga_demo_amcl3_localization/doc/README.md:
```diff
@@ -0,0 +1,95 @@
+
+#### Step 1: File verification
+
+Start the container `ROSDISTRO=jazzy docker/run.sh` and verify that you have the following file structure.
```
@pvela2017 nit: 
```suggestion
Start the container with `ROSDISTRO=jazzy docker/run.sh` and verify that you have the following file structure:
```

pvela2017 replied:
Updated  on [a4e9f6b](https://github.com/Ekumen-OS/beluga-demos/pull/9/commits/a4e9f6b9c8399f41b0280c8c38c07443acad3a15)

hidmic commented on localization/beluga_demo_amcl3_localization/doc/README.md:
```diff
@@ -0,0 +1,95 @@
+
+#### Step 2: Build the demo software code
+
+Now you need to build the demo software stack and run one of the demos.
```
@pvela2017 nit: 
```suggestion
Now you need to build the demo software stack:
```

pvela2017 replied:
Updated  on [a4e9f6b](https://github.com/Ekumen-OS/beluga-demos/pull/9/commits/a4e9f6b9c8399f41b0280c8c38c07443acad3a15)

hidmic commented on localization/beluga_demo_amcl3_localization/doc/README.md:
```diff
@@ -0,0 +1,95 @@
+
+#### Step 4: Run the bag
+
+Finally, run the rosbag:
```
@pvela2017 nit: 
```suggestion
Finally, play the rosbag:
```

pvela2017 replied:
Updated  on [a4e9f6b](https://github.com/Ekumen-OS/beluga-demos/pull/9/commits/a4e9f6b9c8399f41b0280c8c38c07443acad3a15)

hidmic commented on localization/beluga_demo_amcl3_localization/doc/README.md:
````diff
@@ -0,0 +1,95 @@
+
+## Pre-requisites
+
+#### Step 1: Download the VDB map files
+
+The first step is to download the maps into our demo folder:
+
+```bash
+mkdir localization/beluga_demo_amcl3_localization/maps && \
+cd localization/beluga_demo_amcl3_localization/maps && \
+wget url && \
+tar -xf file.tar.gz
+```
+
+#### Step 2: Download the bag files
+
+Then the rosbag files need to be downloaded:
+
+```bash
+mkdir localization/beluga_demo_amcl3_localization/rosbags && \
+cd localization/beluga_demo_amcl3_localization/rosbags && \
+wget url && \
+tar -xf file.tar.gz
+```
````
@pvela2017 meta: it's unlikely we'll expose public endpoints for these but rather make it accessible on-demand. Let's collapse these two as a `Request map and bags from the team` step. CC @glpuga.

pvela2017 replied:
Updated  on [a4e9f6b](https://github.com/Ekumen-OS/beluga-demos/pull/9/commits/a4e9f6b9c8399f41b0280c8c38c07443acad3a15)

pvela2017 commented:
I realized there is a problem when not compiling with the `RELEASE` flag. After one iteration the code crashes in the following [assertion](https://github.com/Ekumen-OS/beluga/blob/ac50ee5b3d42c7462b0103e05737cde4c7f04d23/beluga/include/beluga/algorithm/unscented_transform.hpp#L117C1-L118C1).

```bash
[beluga_amcl3_demo_node-5] [INFO] [1740552394.637537655] [rclcpp]: Particle filter update iteration stats: 1500 particles - 1460.870ms
[beluga_amcl3_demo_node-5] beluga_amcl3_demo_node: /home/developer/ws/install/beluga/include/beluga/beluga/algorithm/unscented_transform.hpp:117: auto beluga::unscented_transform(const Eigen::MatrixBase<Derived>&, const Eigen::MatrixBase<U>&, TransferFn&&, std::optional<typename MeanDerived::Scalar>, MeanFn, ResidualFn) [with MeanDerived = Eigen::Matrix<double, 6, 1, 0, 6, 1>; CovarianceDerived = Eigen::Matrix<double, 6, 6, 0, 6, 6>; TransferFn = Eigen::Matrix<double, 6, 1, 0, 6, 1> (&)(const Eigen::Matrix<double, 6, 1, 0, 6, 1>&); TransformedT = Eigen::Matrix<double, 6, 1, 0, 6, 1>; MeanFn = tf2::toMsg<double>(const Sophus::SE3<double>&, const Eigen::Matrix<double, 6, 6, 0, 6, 6>&)::<lambda(const std::vector<Eigen::Matrix<double, 6, 1, 0, 6, 1>, std::allocator<Eigen::Matrix<double, 6, 1, 0, 6, 1> > >&, const std::vector<double>&)>; ResidualFn = tf2::toMsg<double>(const Sophus::SE3<double>&, const Eigen::Matrix<double, 6, 6, 0, 6, 6>&)::<lambda(Eigen::Vector<double, 6>&, Eigen::Vector<double, 6>&)>; typename MeanDerived::Scalar = double]: Assertion `llt.info() == Eigen::ComputationInfo::Success' failed.
[ERROR] [beluga_amcl3_demo_node-5]: process has died [pid 13376, exit code -6, cmd '/home/developer/ws/install/beluga_demo_amcl3_localization/lib/beluga_demo_amcl3_localization/beluga_amcl3_demo_node --ros-args -r __node:=beluga_amcl3_demo --params-file /tmp/launch_params_zt3d59hn'].
```

hidmic commented:
> I realized there is a problem when not compiling with the RELEASE flag. After one iteration the code crashes in the following [assertion](https://github.com/Ekumen-OS/beluga/blob/ac50ee5b3d42c7462b0103e05737cde4c7f04d23/beluga/include/beluga/algorithm/unscented_transform.hpp#L117C1-L118C1).

@pvela2017 hmm, what does the covariance matrix looks like when not compiling for Release? Is it symmetric positive definite?

pvela2017 commented:
> > I realized there is a problem when not compiling with the RELEASE flag. After one iteration the code crashes in the following [assertion](https://github.com/Ekumen-OS/beluga/blob/ac50ee5b3d42c7462b0103e05737cde4c7f04d23/beluga/include/beluga/algorithm/unscented_transform.hpp#L117C1-L118C1).
> 
> @pvela2017 hmm, what does the covariance matrix looks like when not compiling for Release? Is it symmetric positive definite?



It's symmetric but not positive, here are the values of `covariance` just before the assertion:

 0.24905             0.01235                    0           0           0       0.00271115
 0.01235             0.225493                  0           0           0       0.000583031
 0                         0                                0           0           0           0
 0                         0                                0           0           0           0
 0                         0                                0           0           0           0
 0.00271115       0.000583031           0           0           0       0.0972498

pvela2017 commented:
> > > I realized there is a problem when not compiling with the RELEASE flag. After one iteration the code crashes in the following [assertion](https://github.com/Ekumen-OS/beluga/blob/ac50ee5b3d42c7462b0103e05737cde4c7f04d23/beluga/include/beluga/algorithm/unscented_transform.hpp#L117C1-L118C1).
> > 
> > 
> > @pvela2017 hmm, what does the covariance matrix looks like when not compiling for Release? Is it symmetric positive definite?
> 
> It's symmetric but not positive, here are the values of `covariance` just before the assertion:
> 
> ```
> 0.24905          0.01235           0           0           0       0.00271115
> 0.01235          0.225493          0           0           0       0.000583031
> 0                0                 0           0           0           0
> 0                0                 0           0           0           0
> 0                0                 0           0           0           0
> 0.00271115       0.000583031       0           0           0       0.0972498 
> ```
> 
> Although the covariance matrix is defined as:
> 
> > static constexpr double kInitialPoseCovarianceX = 0.25;
> > static constexpr double kInitialPoseCovarianceY = 0.25;
> > static constexpr double kInitialPoseCovarianceZ = 0.25;
> > static constexpr double kInitialPoseCovarianceYaw = 0.15;
> > static constexpr double kInitialPoseCovariancePitch = 0.15;
> > static constexpr double kInitialPoseCovarianceRoll = 0.15;

I figured out where the problem is!

The problem is the `To3d`  function in the file [3d_embedding.hpp](https://github.com/Ekumen-OS/beluga/blob/ac50ee5b3d42c7462b0103e05737cde4c7f04d23/beluga/include/beluga/3d_embedding.hpp#L31),
```cpp
Sophus::SO3d::rotZ(tf.so2().log()), Eigen::Vector3d{tf.translation().x(), tf.translation().y(), 0.0}};
```
when forcing `roll`, `pitch` and `Z` to 0 the resulting covariance matrix will have those 3 columns and 3 rows of 0 s (corresponding to `Z`, `roll` and `pitch`), making the matrix not symmetric positive definite. I did a quick experiment and added random values and the problem doesnt happens. As shown below:

![image](https://github.com/user-attachments/assets/dfcebf6f-d9d3-44aa-b7ed-b3c54e174f60)


How should we fix this problem?
 - Changing the `To3d` function?
 - Removing the assertion?
 - Compiling just for release?
 - Updating the motion model to 3D? 😏

hidmic commented:
> I figured out where the problem is!

Great find @pvela2017 ! 

> How should we fix this problem?

Hmm, this runs deep. Our embedding "trick" gets us degenerate distributions. I think we have two options:

1. Isolate these tricks in the corresponding models and ensure the filter operates with non-degenerate distributions. In other words, do like we do for the `DifferentialDriveMotionModel` and change the `beluga_vdb::LikelihoodFieldModel3` to accept SE2 states.
2. Communicate that the support of the distribution lies in a lower dimensional space. Can't just pass `Sophus::SE3` objects, we would need to pass a special `Sophus::SE3`* to tell `tf2::toMsg` it has to project it before attempting an unscented transform.

Personally, I think 1. is much less cumbersome. CC'ing @glpuga and @nahueespinosa for feedback. This is broken for NDT too guys.

glpuga commented:
> > It's symmetric but not positive, here are the values of `covariance` just before the assertion:
> > ```
> > 0.24905          0.01235           0           0           0       0.00271115
> > 0.01235          0.225493          0           0           0       0.000583031
> > 0                0                 0           0           0           0
> > 0                0                 0           0           0           0
> > 0                0                 0           0           0           0
> > 0.00271115       0.000583031       0           0           0       0.0972498 
> > ```

Totally makes sense, since this matrix has half of the eigen values in zero. It's positive semi-definite, and standard Cholesky has no power here.

Unfortunately I have zero understanding of https://github.com/Ekumen-OS/beluga/blob/ac50ee5b3d42c7462b0103e05737cde4c7f04d23/beluga/include/beluga/algorithm/unscented_transform.hpp#L86 and have no idea where that algorithm came from, but the Cholesky decomposition still can be applied on semi-definite matrices, but may not be unique (unless a pivot is chosen). 

https://en.wikipedia.org/wiki/Cholesky_decomposition#Positive_semidefinite_matrices
https://eigen.tuxfamily.org/dox/classEigen_1_1LDLT.html

Whether that is helpful here or not, I don't really know.

Regardless, I think having 2d/3d versions of this is the way to go.

pvela2017 commented:
> > > It's symmetric but not positive, here are the values of `covariance` just before the assertion:
> > > ```
> > > 0.24905          0.01235           0           0           0       0.00271115
> > > 0.01235          0.225493          0           0           0       0.000583031
> > > 0                0                 0           0           0           0
> > > 0                0                 0           0           0           0
> > > 0                0                 0           0           0           0
> > > 0.00271115       0.000583031       0           0           0       0.0972498 
> > > ```
> 
> Totally makes sense, since this matrix has half of the eigen values in zero. It's positive semi-definite, and standard Cholesky has no power here.
> 
> Unfortunately I have zero understanding of https://github.com/Ekumen-OS/beluga/blob/ac50ee5b3d42c7462b0103e05737cde4c7f04d23/beluga/include/beluga/algorithm/unscented_transform.hpp#L86 and have no idea where that algorithm came from, but the Cholesky decomposition still can be applied on semi-definite matrices, but may not be unique (unless a pivot is chosen).
> 
> https://en.wikipedia.org/wiki/Cholesky_decomposition#Positive_semidefinite_matrices https://eigen.tuxfamily.org/dox/classEigen_1_1LDLT.html
> 
> Whether that is helpful here or not, I don't really know.
> 
> Regardless, I think having 2d/3d versions of this is the way to go.

So, should I open a new issue in beluga repo and implement the 2d/3d version?

What about this PR, should we merge it and then update it?

Also, Im waiting for [beluga #475](https://github.com/Ekumen-OS/beluga/pull/475) to be merged to rollback the distro specific switches.

> @pvela2017 meta: if you are up for [Ekumen-OS/beluga#474](https://github.com/Ekumen-OS/beluga/issues/474) once we've sorted out the OpenVDB dependency in Humble, we should come back here and rollback the distro specific switches.

hidmic commented:
> So, should I open a new issue in beluga repo and implement the 2d/3d version?

Yes please!

> What about this PR, should we merge it and then update it?

Hmm, I'm don't love the idea of merging broken code.

> Also, Im waiting for https://github.com/Ekumen-OS/beluga/pull/475 to be merged to rollback the distro specific switches.

Makes sense.

pvela2017 commented:
> > So, should I open a new issue in beluga repo and implement the 2d/3d version?
> 
> Yes please!
> 
> > What about this PR, should we merge it and then update it?
> 
> Hmm, I'm don't love the idea of merging broken code.
> 
> > Also, Im waiting for [Ekumen-OS/beluga#475](https://github.com/Ekumen-OS/beluga/pull/475) to be merged to rollback the distro specific switches.
> 
> Makes sense.

Issue created, please check it: https://github.com/Ekumen-OS/beluga/issues/478

pvela2017 commented:
I was testing the demo in Humble, but there are a few small problems:

1.- The PCL filters (CropBox and VoxelGrid, actually all the filters) are not [compiled as a plugin](https://github.com/ros-perception/perception_pcl/blob/humble/pcl_ros/CMakeLists.txt), thus they can't be run as a [node](https://github.com/pvela2017/beluga-demos/blob/61f59b2c7dd196ebfc0a8b583c10240446b12c45/localization/beluga_demo_amcl3_localization/launch/demo_botanic_garden_amcl3_localization.launch.py#L54).  I don't know how long each iteration will take without the filters (to decrease the number of points of the pointcloud).

2.- The Bag was recorded on Jazzy, so it can't be played on Humble.

Should we re-record the bag using Humble and create some scripts to run the filters? Or is enough that the demo compiles on Humble and it runs only on Jazzy?

pvela2017 commented:
> Let me reverse that question. What do you think we should do?

mmmmm... ~~We should focus on Jazzy and tell the new intern to fix the issues on Humble.~~ :laughing:
Well we did put efforts to make `beluga_vdb` available on Humble so it makes sense to fix the demo and if I give a second thought I don't think it would be that much work. But there is still a lot of work to do to make a 2D version of `beluga_vdb` so I would like to focus on that first. 

So the roadmap for me would be:
```
Implement the 2D version of `beluga_vdb` --> Adapt the demo for Jazzy --> Fix demo for Humble
```

Also, I would to implement the 3D version of the motion model :grin:

> Remove the meta file, and run ros2 bag reindex on the bagfile.

Thank you! I just tried that but it didn't work. I guess because the bag is in `mcap` format. I will try to transform it to `sqlite3` first and then reindex it.

```sh
sudo apt install ros-humble-rosbag2-storage-mcap
ros2 bag reindex ./botanic_garden_tf_1005_07 -s mcap
[INFO] [1741858100.467437013] [rosbag2_cpp]: Beginning reindexing bag in directory: ./botanic_garden_tf_1005_07
[INFO] [1741858100.482302943] [rosbag2_cpp]: Reindexing complete.
```
```sh
ros2 bag play botanic_garden_tf_1005_07/botanic_garden_tf_1005_07.mcap 

closing.
[INFO] [1741858113.569076048] [rosbag2_player]: Set rate to 1
yaml-cpp: error at line 1, column 12: bad conversion

```

glpuga commented:
> Thank you! I just tried that but it didn't work. I guess because the bag is in mcap format. I will try to transform it to sqlite3 first and then reindex it.

Sounds familiar. Also, looks like the same issue discussed here: https://github.com/ros2/rosbag2/issues/1895#issuecomment-2680358501 There's still hope.

pvela2017 commented:
Sorry! I have been a little busy recently.

I updated the demo according to the [Beluga #PR480](https://github.com/Ekumen-OS/beluga/pull/480), but there is a huge degradation.

![image](https://github.com/user-attachments/assets/c6438d64-888c-45c8-930b-66f0aaf865e5)

pvela2017 commented:
> @hidmic Here are the new rosbags https://drive.google.com/drive/folders/1BAa4U35xnPMPY7RjMHsCF_-uWxES5HHB?usp=sharing

This link now contains the update `.db3` rosbags that can be played on Humble and Jazzy.

pvela2017 commented:
I hid all the problems inside the humble `Dockerfile`. Now the jazzy version of the pcl_perception repo is installed in the humble image when it is built. I think this is the cleanest solution. The other solution I thought was to copy the files in the demo and modify the CMake and launcher.

Everything works well on Jazzy. But for Humble, if the bag is played with `--clock` the following problem happens:

```sh
ros2 bag play bag.db3 --clock
[filter_crop_box_node-4] [WARN] [1745076261.603625802] [tf2_buffer]: Detected jump back in time. Clearing TF buffer. 
[filter_voxel_grid_node-3] [WARN] [1745076261.603713806] [tf2_buffer]: Detected jump back in time. Clearing TF buffer.
[beluga_amcl3_demo_node-5] [INFO] [1745076261.609829635] [beluga_amcl3_demo]: Message Filter dropping message: frame 'velodyne' at time 1664954581.391 for reason 'the timestamp on the message is earlier than all the data in the transform cache'
```

I dont know why this is an issue on Humble and not in Jazzy, but since it works without `--clock` for both I just updated the documentation and remove the `--clock` arg.

hidmic commented on localization/beluga_demo_amcl3_localization/launch/demo_botanic_garden_amcl3_localization.launch.py:
```diff
@@ -0,0 +1,111 @@
+        parameters=[{'use_sim_time': True, 'leaf_size': 0.1}],
+        remappings=[
+            ('/input', '/velodyne_points'),
+            # ('/output', '/pointcloud'),
```
@pvela2017 nit: drop commented out code

pvela2017 replied:
Fixed on [0586827](https://github.com/Ekumen-OS/beluga-demos/pull/9/commits/0586827ca88e9a9f92bc0435ad95fe35c6a0e781)

hidmic commented on localization/beluga_demo_amcl3_localization/src/beluga_amcl3_demo.cpp:
```diff
@@ -0,0 +1,511 @@
+
+    // Callback options
+    callback_group_ =
+        this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
```
@pvela2017 hmm, reentrancy means that timer and pointcloud callback may run in parallel if the node is dispatched to a multi-threaded executor. There's no lock synchronizing access to the underlying particle filter. Should it be a mutually exclusive callback group instead?

pvela2017 replied:
Right! my bad fixed on [8d53352](https://github.com/Ekumen-OS/beluga-demos/pull/9/commits/8d53352f1cc531eeffa50d1672c4d460c687848d)

hidmic reviewed (approved):
Amazing job!

hidmic commented:
@pvela2017 which level set resolution were the above run against?

pvela2017 commented:
> @pvela2017 which level set resolution were the above run against?

For the [VDB Mapping part](https://github.com/pvela2017/beluga/blob/vdb_mapping/vdb_mapping_ros2/vdb_mapping_ros2/config/vdb_params.yaml) the resolution is 7 cm and the threshold max is  0.51.
Then for the [code](https://github.com/pvela2017/beluga/blob/vdb_mapping/vdb_mapping_to_levelset/src/narrowband.cpp) that transform the map to a level set the `halfWidth` parameter was set to 8 in [`topologyToLevelSet()`](https://www.openvdb.org/documentation/doxygen/namespaceopenvdb_1_1v12__0_1_1tools.html#aabbae85c9cdcbfc824ca7222995f16ca)

hidmic reviewed (approved):
Wonderful
